modules/publish/Components.php

```diff
@@ -132,4 +134,162 @@ public function componentSpecialWhatLinksHere()
+                                             'revision' => $revision,
+                                             'article_name' => $article_name,
+                                             'reason' => $reason,
+                                             'author' => $author,
+                                             'author_contributions_url' => $author_contributions_url,
+                                             'revision_url' => $revision_url,
+                                             'article_url' => $article_url,
+                                             'diff_url' => $diff_url,
+                                             'history_url' => $history_url);
+                }
+            }
+
+            // Ensure the page size is a valid value (has to be whole number greated than 0).
+            if ($page_size < 1 )
+            {
+                $page_size = 1;
+            }
+
+            // Calculate number of contributions and how many pages we have.
+            $contributions_size = sizeof($contributions);
+            $total_pages = ceil($contributions_size / $page_size);
+
+            // Ensure we are not out of bounds with page number.
+            if ($page > $total_pages)
+            {
+                $page = $total_pages;
+            }
+            elseif ($page < 1)
+            {
+                $page = 1;
+            }
+
+            // Determine starting and ending contribution that fits on this specific page.
+            $page_start_at = ($page - 1) * $page_size;
+            $page_end_at = ($page - 1) * $page_size + $page_size - 1;
+
+            // Make sure we don't go past the last contribution.
+            if ($page_end_at > $contributions_size - 1)
+            {
+                $page_end_at = $contributions_size - 1;
+            }
+
+            // Calculate pagination URLs.
+            $navigation_urls = array();
+            $page_size_urls = array();
+
+            $navigation_url_base = $base_url . $base_url_user_prefix . ($page_size != $default_page_size ? 'page_size=' . $page_size : '');
+
+            $navigation_urls['newest'] = $navigation_url_base . '&page=' . '1';
+            $navigation_urls['oldest'] = $navigation_url_base . '&page=' . $total_pages;
+            $navigation_urls['newer']  = $navigation_url_base . '&page=' . ($page > 1 ? $page - 1 : 1);
+            $navigation_urls['older']  = $navigation_url_base . '&page=' . ($page < $total_pages ? $page + 1 : $total_pages);
+
+            foreach ($available_page_sizes as $available_page_size)
+            {
+                $page_size_urls[$available_page_size] = $base_url . $base_url_user_prefix . 'page_size='  . $available_page_size;
+            }
+
+            // Prepare context for template.
+            $this->username = $username;
+            $this->user = $user;
+            $this->contributions = $contributions;
+
+            $this->page = $page;
```

Comment:
I can see how this would be a useful pattern, but it might be a bit over my head. Couple of questions:

1. Should we then switch to using same pagination style everywhere. E.g. new publish code does it MediaWiki style with the older/newer links, while search pagination does it with numbers (1 2 3 4...100). Probably would be best to use number-style.

2. How would one render this on a page - i.e. how and where I should put the code so it can behave a bit like a mix-in?

3. This would touch on some rather core functionality, and I'm a bit afraid to increase the scope of this pull request. How about a separate pull request after this one gets merged (knowing that I can't guarantee exactly when, but I'll try :)?